src/data/data.cc

```diff
@@ -40,10 +40,35 @@ void MetaInfo::Clear() {
 }
 
 void MetaInfo::SaveBinary(dmlc::Stream *fo) const {
+  /**
+   * Header: version, num_field, followed by list of form [ (field_name, field_offset) ].
+   * Offset is to be calculated from the end of the header.
+   **/
   Version::Save(fo);
-  fo->Write(&num_row_, sizeof(num_row_));
-  fo->Write(&num_col_, sizeof(num_col_));
-  fo->Write(&num_nonzero_, sizeof(num_nonzero_));
+  const uint64_t num_field = kNumField;
+  fo->Write(num_field);
+  uint64_t offset = 0;
+  int field_cnt = 0;  // make sure we are actually writing kNumField fields
+  fo->Write(std::string(u8"num_row")); fo->Write(offset);
+  offset += sizeof(num_row_); ++field_cnt;
+  fo->Write(std::string(u8"num_col")); fo->Write(offset);
+  offset += sizeof(num_col_); ++field_cnt;
+  fo->Write(std::string(u8"num_nonzero")); fo->Write(offset);
+  offset += sizeof(num_nonzero_); ++field_cnt;
+  fo->Write(std::string(u8"labels")); fo->Write(offset);
+  offset += sizeof(decltype(labels_)::value_type) * labels_.Size(); ++field_cnt;
+  fo->Write(std::string(u8"group_ptr")); fo->Write(offset);
+  offset += sizeof(decltype(group_ptr_)::value_type) * group_ptr_.size(); ++field_cnt;
+  fo->Write(std::string(u8"weights")); fo->Write(offset);
+  offset += sizeof(decltype(weights_)::value_type) * weights_.Size(); ++field_cnt;
+  fo->Write(std::string(u8"base_margin")); fo->Write(offset);
+  offset += sizeof(decltype(base_margin_)::value_type) * base_margin_.Size(); ++field_cnt;
+  CHECK(field_cnt == kNumField) << "Wrong number of fields";
```

Comment:
Check eq.

Reply:
Apparently, the compiler is not allocating memory for the static constant `kNumField`, so `CHECK_EQ` results in a linking error.

Reply:
You can allocate it outside of class, see the static fields in learner.  But as this works so I'm OK with normal CHECK.

Reply:
Ah got it. Let me do that.